be/src/exec/pipeline/exchange/shuffler.h

```diff
@@ -0,0 +1,93 @@
+        }
+
+        if (_part_type == TPartitionType::HASH_PARTITIONED && !compatibility) {
+            _local_exchange_shuffle = &Shuffler::local_exchange_shuffle<ReduceOp>;
+        } else {
+            _local_exchange_shuffle = &Shuffler::local_exchange_shuffle<ModuloOp>;
+        }
+    }
+
+    void exchange_shuffle(std::vector<uint32_t>& shuffle_channel_ids, const std::vector<uint32_t>& hash_values,
+                          size_t num_rows) {
+        (this->*_exchange_shuffle)(shuffle_channel_ids, hash_values, num_rows);
+    }
+
+    void local_exchange_shuffle(std::vector<uint32_t>& shuffle_channel_ids, std::vector<uint32_t>& hash_values,
+                                size_t num_rows) {
+        (this->*_local_exchange_shuffle)(shuffle_channel_ids, hash_values, num_rows);
+    }
+
+private:
+    void (Shuffler::*_exchange_shuffle)(std::vector<uint32_t>& shuffle_channel_ids,
+                                        const std::vector<uint32_t>& hash_values, size_t num_rows) = nullptr;
+
+    void (Shuffler::*_local_exchange_shuffle)(std::vector<uint32_t>& shuffle_channel_ids,
+                                              std::vector<uint32_t>& hash_values, size_t num_rows) = nullptr;
+
+    template <bool two_level_shuffle, typename ReduceOp>
+    void exchange_shuffle(std::vector<uint32_t>& shuffle_channel_ids, const std::vector<uint32_t>& hash_values,
+                          size_t num_rows) {
+        for (size_t i = 0; i < num_rows; ++i) {
+            size_t channel_id = ReduceOp()(hash_values[i], _num_channels);
+            size_t shuffle_id;
+            if constexpr (!two_level_shuffle) {
+                shuffle_id = channel_id;
+            } else {
+                uint32_t driver_sequence = ReduceOp()(HashUtil::xorshift32(hash_values[i]), _num_shuffles_per_channel);
+                shuffle_id = channel_id * _num_shuffles_per_channel + driver_sequence;
+            }
+            shuffle_channel_ids[i] = shuffle_id;
+        }
+    }
+
+    // When the local exchange is the successor of ExchangeSourceOperator,
+    // the data flow is `exchange sink -> exchange source -> local exchange`.
+    // `exchange sink -> exchange source` (phase 1) determines which fragment instance to deliver each row,
+    // while `exchange source -> local exchange` (phase 2) determines which pipeline driver to deliver each row.
+    // To avoid hash two times and data skew due to hash one time, phase 1 hashes one time
+    // and phase 2 applies xorshift32 on the hash value.
+    // Note that xorshift32 rehash must be applied for both local shuffle here and exchange sink.
+    template <typename ReduceOp>
+    void local_exchange_shuffle(std::vector<uint32_t>& shuffle_channel_ids, std::vector<uint32_t>& hash_values,
+                                size_t num_rows) {
+        for (int32_t i = 0; i < num_rows; ++i) {
+            uint32_t driver_sequence = ReduceOp()(HashUtil::xorshift32(hash_values[i]), _num_shuffles_per_channel);
```

Comment:
Why use `_num_shuffles_per_channel` not `_num_channels`? Maybe `_num_shuffles_per_channel` is only used by `two_level_shuffle`?

Reply:
sounds reasonable

Reply:
fixed